hypershift-operator/controllers/resourcebasedcpautoscaler/controller.go

```diff
@@ -0,0 +1,406 @@
+	controller := &ControlPlaneAutoscalerController{
+		Client: mgr.GetClient(),
+	}
+	if err := ctrl.NewControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(&hyperv1.HostedCluster{}).
+		Watches(&vpaautoscalingv1.VerticalPodAutoscaler{}, handler.EnqueueRequestsFromMapFunc(hostedClusterForVPA)).
+		Complete(controller); err != nil {
+		return err
+	}
+	return nil
+}
+
+func hostedClusterForVPA(ctx context.Context, object client.Object) []reconcile.Request {
+	vpa, isVPA := object.(*vpaautoscalingv1.VerticalPodAutoscaler)
+	if !isVPA {
+		return nil
+	}
+	ns := vpa.Annotations[hcNamespaceAnnotation]
+	name := vpa.Annotations[hcNameAnnotation]
+	if ns == "" || name == "" {
+		return nil
+	}
+	return []reconcile.Request{
+		{
+			NamespacedName: types.NamespacedName{
+				Namespace: ns,
+				Name:      name,
+			},
+		},
+	}
+}
+
+func (r *ControlPlaneAutoscalerController) updateSizeCache(ctx context.Context) error {
+	if r.updateSizeCacheFunc != nil {
+		return r.updateSizeCacheFunc(ctx)
+	}
+	config := &schedulingv1alpha1.ClusterSizingConfiguration{}
+	config.Name = "cluster"
+	if err := r.Get(ctx, client.ObjectKeyFromObject(config), config); err != nil {
+		return fmt.Errorf("failed to fetch cluster sizing configuration: %w", err)
+	}
+
+	listMachineSets := func() (*machinev1beta1.MachineSetList, error) {
+		result := &machinev1beta1.MachineSetList{}
+		err := r.List(ctx, result, client.InNamespace("openshift-machine-api"))
+		return result, err
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+	if err := r.sizeCache.update(config, listMachineSets, log); err != nil {
+		return fmt.Errorf("failed to update machine size cache: %w", err)
+	}
+	return nil
+}
+
+func (r *ControlPlaneAutoscalerController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	if err := r.updateSizeCache(ctx); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	hc := &hyperv1.HostedCluster{}
+	if err := r.Get(ctx, request.NamespacedName, hc); err != nil {
+		if apierrors.IsNotFound(err) {
+			// HostedCluster was deleted or does not exist; nothing to do
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if !hc.DeletionTimestamp.IsZero() {
+		// No need to process deleting HostedClusters
+		return ctrl.Result{}, nil
+	}
+
+	cpNamespace := manifests.HostedControlPlaneNamespace(hc.Namespace, hc.Name)
+	vpa := controlplaneautoscalermanifests.KubeAPIServerVerticalPodAutoscaler(cpNamespace)
+
+	if hc.Annotations[hyperv1.ResourceBasedControlPlaneAutoscalingAnnotation] != "true" || hc.Annotations[hyperv1.TopologyAnnotation] != hyperv1.DedicatedRequestServingComponentsTopology {
+		_, err := util.DeleteIfNeeded(ctx, r, vpa)
+		return ctrl.Result{}, err
+	}
+
+	// Ensure VPA is up to date
+	if _, err := controllerutil.CreateOrUpdate(ctx, r.Client, vpa, reconcileKubeAPIServerVPAFunc(vpa, hc)); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to reconcile kube apiserver vpa: %w", err)
+	}
+
+	if !isVPAReady(vpa) {
```

Comment:
what if VPA will never be ready? I would add a condition eventually for metric + alert. But a follow-up PR is more than OK